src/Chainweb/Cut.hs

```diff
@@ -305,48 +401,86 @@ instance Exception CutException
--- cut.
+-- This check is only needed for external cuts. Note, that for imported cuts
+-- this must be checked recursively.
 --
-checkBraidingOfCut
-    :: MonadThrow m
-    => Cut
-    -> m ()
+-- The implementation of this check here is expensive, namely \(O(m)\), where
+-- \(m\) is the size of the chain graph, i.e. the order of the graph times the
+-- degree. Usually, a more efficient way to check this property for a new cut is
+-- to performm a join that starts with a meet with a local cut, where the local
+-- cut is know to have this property.
+--
+-- Thus, the functions in this section are mostly used for assertions in tests.
+-- They also serve as reference implementations and documentation.
+--
+checkBraidingOfCut :: MonadThrow m => Cut -> m ()
 checkBraidingOfCut = checkBraidingOfCutPairs . _cutAdjPairs
+{-# INLINE checkBraidingOfCut #-}
 
 -- | Check that a set of adjacent pairs of a cut is correctly braided.
 --
 checkBraidingOfCutPairs
     :: MonadThrow m
     => Foldable f
-    => f (AdjPair BlockHeader)
+    => f (BlockHeader, BlockHeader)
     -> m ()
-checkBraidingOfCutPairs = traverse_ checkBraidingOfCutPair
+checkBraidingOfCutPairs = traverse_ (uncurry checkBraidingOfCutPair)
+{-# INLINE checkBraidingOfCutPairs #-}
 
+-- | Checks that directed adjacent pair in a cut is correctly braided.
+--
 checkBraidingOfCutPair
     :: MonadThrow m
-    => AdjPair BlockHeader
+    => BlockHeader
+        -- ^ Source header
+    -> BlockHeader
+        -- ^ target header
     -> m ()
-checkBraidingOfCutPair p = unlessM (isBraidingOfCutPair p)
-    $ throwM (InvalidCutPair p)
+checkBraidingOfCutPair s t = do
+    unless (absBlockHeightDiff s t <= 1) $ throwM $ InvalidCutPair s t
+    unlessM (isBraidingOfCutPair s t) $ throwM (InvalidCutPair s t)
 
--- | Returns whether an adjacent pair in a cut is correctly braided.
+-- | Returns whether a directed adjacent pair in a cut is correctly braided.
 --
 -- * throws 'ChainNotAdjacentException'
 --
+-- Assuming that all headers in the cut have the same graph, the following
+-- condition for all undirected adjacent pairs in the cut would inductively
+-- guarantee a valid braiding:
+--
+-- @
+-- isBraidingOfCutPair (AdjPair a b) = do
+--    ab <- getAdjacentHash b a -- adjacent of a on chain of b
+--    ba <- getAdjacentHash a b -- adajcent of b on chain of a
+--    return
+--        $! (_blockParent a == ba && _blockParent b == ab)
+--        || ab == _blockHash b
+--        || ba == _blockHash a
+-- @
+--
+-- The actual implementation is a it more complex because headers of different
+-- height in a cut may use a diferent chain graph during the transition from one
+-- graph to another graph: instead of undirected adjacent pairs the condition
+-- consideres directed adjacent pairs.
+--
+-- A corresponding function is implemente in 'Chainweb.Miner.Coordinator'
+-- TODO: provide test that the implementation in the miner matches with this
+-- function.
```

Comment:
I couldn't find this in `Coordinator`, is this comment still true?

Reply:
It's not true any more. I updated the comment.